### ISSUE TITLE
Annotate records and primary constructors with marker attributes

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#45777]
Currently there's no way to tell if a type is a record from metadata. I think this would be useful for a variety of purposes:

e.g. 
- MetadataAsSource/decompilers could more easily show records as source
- I was discussing with someone who wanted to make all records implement GetEnumerator over the fields via an extension method using a source generator. Currently you would have to detect records from metadata by convention.

To do so I think it would be useful to have an attribute marking a Record, and another one marking the primary constructor of a positional record.

**Comment by CyrusNajmabadi** [comment]
I believe there's an unspeakable 'Clone' method you can look for that will tell you if its a record.

**Comment by RikkiGibson** [comment]
I agree that it could be useful to be able to identify the record primary constructor from metadata. I am not aware of any way to do this.

**Comment by YairHalberstadt** [comment]
It looks like Asp.Net could do with this: https://github.com/dotnet/aspnetcore/issues/23975

**Comment by YairHalberstadt** [comment]
If the  team plans to do mark primary constructors with a special attribute, it would be best if they do it before C# 9.0 is released, so that all primary constructors will have this attribute. Otherwise it will only workfor some records, depending on when it was compiled.

**Comment by Andrei15193** [comment]
Hi.

Has there been any update on this? If not, it would be useful to know how the compiler distinguishes record types from class types. If you try to inherit a class from a record, e.g.: `record MyRecord : Stream`, the compiler gives the following error: `error CS8864: Records may only inherit from object or another record`. It definitely knows how to determine that.

I'm working on a project to generate documentation out of assemblies and related generated XML documents (the documentation comments you add in source code, e.g.: `/// <summary> ... </summary>`). Knowing what record types a library exposes would be really useful as it will provide clear documentation in that sense.

I'm currently checking for the unspeakable 'Clone' method to determine this, but knowing whether this will remain valid long-term would help me a great deal. Thanks!
